pennylane/ops/qubit.py

```diff
@@ -3087,6 +3088,41 @@ def diagonalizing_gates(self):
         return []
 
 
+class SparseHamiltonian(Observable):
+    r"""SparseHamiltonian(H)
+    A Hamiltonian represented directly as a sparse matrix in coordinate list (COO) format.
+
+    **Details:**
+
+    * Number of wires: Any
+    * Number of parameters: 1
+    * Gradient recipe: None
+
+    Args:
+        H (coo_matrix): a sparse matrix in scipy coordinate list (COO) format with
+            dimension :math:`(2^n, 2^n)`, where :math:`n` is the number of wires
+    """
+    num_wires = AnyWires
+    num_params = 1
+    par_domain = None
+    grad_method = None
+
+    wires = Wires([0])
+
+    def __init__(self, *params, wires=wires):
+        super().__init__(*params, wires=wires)
```

Comment:
@soranjh it just occurs to me that this requires that the sparse Hamiltonian must be applied to all wires on the device.

However, can we not also allow expectation values of sparse hamiltonians on a _subset_ of wires?

```python
H = scipy.sparse.eye(4).tocoo()

dev = qml.device("default.qubit", wires=5)

@qml.qnode(dev)
def circuit():
    ...
    return qml.expval(qml.SparseHamiltonian(H, wires=[0, 2]))
```

This feels like behaviour that would both be (a) useful, and (b) intuitive to users based on how PennyLane currently works --- for example, it matches `qml.Hermitian`.

Reply:
Hey @josh146. What you're suggesting is indeed possible, but would require significant additional work, while targeting a situation that should essentially never occur in practice. Let me explain.

Basically, the way that expvals are currently computed over a subset of wires is to compute marginal probabilities, then use eigenvalues of local observables to compute expectation values. We can't use that logic, since it would require us to diagonalize the full Hamiltonian. Instead, I see two ways to support this:

1. Compute identity\otimes H, then calculate the expval wrt to the entire state. This then gives the expval over the reduced state on a subset of wires. 
2. Calculate the reduced density matrix over the subset of wires, then compute Tr(\rho H).

So supporting this would require us to essentially change the entire PR. 

But the biggest motivation not to worry about supporting this is that, at least for all quantum chemistry applications, we always want to prepare a *pure state* related to the Hamiltonian, typically an eigenstate. Computing an expval over a subset of wires would effectively mean the other wires are used to prepare a mixed state over the subset, which simply appears as noise in the state; something we want to avoid. So in all cases we want the number of wires in the Hamiltonian and the device to be the same.

Maybe let's wait to see if users ever actually ask for supporting expvals of sparse Hamiltonians over a subset of wires before considering supporting that

Reply:
:+1:

In that case, I propose raising an error if `wires` is passed by the user to `SparseHamiltonian`, to ensure that do not get confused by its functionality 🙂

Reply:
@josh146 PL raises an error if no wire is specified, that is why I chose to have a default of [0]. Now that I am thinking more, it seems to me that using `num_wires = AllWires` in the class makes more sense (conceptually the sparse observable is applied to all wires). In this case, the user should pass all wires otherwise gets an error saying that all wires must be passed. If we chose to not have any wires at all or pass an empty list, then PL raises another error. I tried couple of other tricks but all failed some tests. So my suggestion is to use `num_wires = AllWires`, pass all wires and everything will be fine. Any thoughts?

Reply:
@soranjh you are exactly right -- sorry, `AllWires` skipped my mind previously, but this is exactly what is intended to be used for.